lib/AuthStateManager.ts

```diff
@@ -42,22 +43,23 @@ class AuthStateManager {
     this._sdk = sdk;
     this._pending = { ...DEFAULT_PENDING };
     this._authState = { ...DEFAULT_AUTH_STATE };
+    this._lastEventTimestamp = 0;
 
     // Listen on tokenManager events to start updateState process
     // "added" event is emitted in both add and renew process, just listen on "added" event to update auth state
-    sdk.tokenManager.on(EVENT_ADDED, (key, token) => {
-      this.updateAuthState({ event: EVENT_ADDED, key, token });
+    sdk.tokenManager.on(EVENT_ADDED, (key, token, { timestamp }) => {
+      this.updateAuthState({ event: EVENT_ADDED, key, token, timestamp });
     });
-    sdk.tokenManager.on(EVENT_REMOVED, (key, token) => {
-      this.updateAuthState({ event: EVENT_REMOVED, key, token });
+    sdk.tokenManager.on(EVENT_REMOVED, (key, token, { timestamp }) => {
+      this.updateAuthState({ event: EVENT_REMOVED, key, token, timestamp });
     });
   }
 
   getAuthState(): AuthState {
     return this._authState;
   }
 
-  updateAuthState({ event, key, token }: UpdateAuthStateOptions = {}): void {
+  updateAuthState({ event, key, token, timestamp }: UpdateAuthStateOptions = { timestamp: Date.now() }): void {
```

Comment:
I do not think this function should take any parameters or options. None of these parameters should affect the output. Is the output state a function of an event or a key? All of these params (except for timestamp) appear to be here for logging purposes. I would prefer this type of logging was done outside of this method. If an event handler has been fired, that handler can log a message before calling `updateAuthState`. `updateAuthState` itself should not take any parameters or return any value, since it should calculate the same state no matter who calls it nor however frequently it is called. Two identical calls in a row should produce the same output state. 

Adding time-based logic with timestamp and Date.now() greatly increases the complexity and breaks the assumption of repeatability / idempotence since two subsequent calls with the same parameters would have different results. With the time logic, each call causes an update to internal state.  

Calling this method should not produce side-effects other than emitting a state change event with the calculated state object, which is a function of stored token values. It should not matter when I call the method, the current state will always be a reflection of current storage

Reply:
I think we can still track the `timestamp`, but instead of tracking it in `updateAuthState`, we can manage it when AuthStateManager gets the events, then make the decision there. 
In this way, the internal evaluation can still benefit from the tracked timestamp, and the external evaluation won't break the assumption of repeatability/idempotence.